projects/igniteui-angular/src/lib/services/overlay/overlay.ts

```diff
@@ -735,6 +726,22 @@ export class IgxOverlayService implements OnDestroy {
         }
     }
 
+    private addCloseOnEscapeListener() {
+        //  if all overlays minus closing overlays equals one add the handler
+        if (this._overlayInfos.length - this._overlayInfos.filter(x => x.closeAnimationPlayer
+            && x.closeAnimationPlayer.hasStarted()).length === 1) {
+            fromEvent(this._document, 'keydown').pipe(
```

Comment:
Right, so once this subscription is active, there's no mechanism to remove it (outside of destroy, which is basically never for a root service)? I don't like a key handler to be permanently active TBH. Also the check is all wrong - length alone doesn't tell you much, where's the `closeOnEsc` before setting the listener? And won't this keep adding subscriptions if just one overlay is shown at a time (length === 1)?

Don't get me wrong, the old version with subscribing to the wrapper is also no good as it never unsubscribed and rxjs won't do that for removed elements either. If you want to use the event observable - do store subscription, check `closeOnEsc` before even setting it up and use similar logic to unsubscribe like outside click.

Reply:
Unsubscribing is fixed now.
This will never subscribe more than once. When calling `show` for the first overlay it will subscribe. Then if you close this overlay and during its close animation open a new one two things will happen. First in `hide` call subscription will be removed. Then in `show` there will be still two overlays active, and one with closing animation playing - it will subscribe which is correct.